### PR TITLE
Adding support for obsm, varm, obsp, varp

### PR DESCRIPTION
**Pull request by ivirshup**
Closes: #20 #21 #22 #23

Currently adds ability to hold `obsm` + `varm` to `Abstract`, `HDF5`, and `InMemory`.

Todo:

- [x] Test IO
- [x] Validation
  - [x] InMemory
  - [x] H5AD
- [x] Docs

I think I would like to leave `SingleCellExperiment` +  `Seurat` to separate PRs.

### CODE REVIEW

File: R/AbstractAnnData.R [hunk 1]
```diff
@@ -155,6 +163,64 @@ AbstractAnnData <- R6::R6Class("AbstractAnnData", # nolint
...
+    },
+
+    .validate_array_collection_generic = function(collection, label, shape, expected_rownames=NULL, expected_colnames=NULL) {
+      if (is.null(collection)) {
+        return(collection)
+      }
+
+      collection_names <- names(collection)
+      if (!is.list(collection) || is.null(collection_names)) {
+        stop("collection must must be a named list")
```
Comment (rcannood) [inline R/AbstractAnnData.R:213]: Add label to error message

```suggestion
        stop(paste0(label, " must must be a named list"))
```

File: R/InMemoryAnnData.R [hunk 2]
```diff
@@ -118,6 +120,25 @@ InMemoryAnnData <- R6::R6Class("InMemoryAnnData", # nolint
         private$.var_names <- private$.validate_obsvar_names(value, "var")
         self
       }
+    },
+    #' @field obsm
+    obsm = function(value) {
+      if (missing(value)) {
+        private$.obsm
```
Comment (rcannood) [inline R/InMemoryAnnData.R:127]: Add trackstatus comment

```suggestion
        # trackstatus: class=InMemoryAnnData, feature=get_obsm, status=done
        private$.obsm
```

File: R/InMemoryAnnData.R [hunk 3]
```diff
@@ -118,6 +120,25 @@ InMemoryAnnData <- R6::R6Class("InMemoryAnnData", # nolint
...
+      } else {
+        # TODO: validate obsm
```
Comment (rcannood) [inline R/InMemoryAnnData.R:129]: Add trackstatus comment

```suggestion
        # trackstatus: class=InMemoryAnnData, feature=set_obsm, status=wip
        # TODO: validate obsm
```

File: R/InMemoryAnnData.R [hunk 4]
```diff
@@ -118,6 +120,25 @@ InMemoryAnnData <- R6::R6Class("InMemoryAnnData", # nolint
...
+        private$.obsm <- private$.validate_array_collection_generic(value, "obsm", c(self$n_obs()), expected_rownames=rownames(self))
+        self
+      }
+    },
+    varm = function(value) {
+      if (missing(value)) {
+        private$.varm
+      } else {
+        # TODO validate varm
```
Comment (rcannood) [inline R/InMemoryAnnData.R:138]: Add trackstatus comment

```suggestion
        # trackstatus: class=InMemoryAnnData, feature=set_varm, status=wip
        # TODO: validate varm
```

File: R/InMemoryAnnData.R [hunk 5]
```diff
@@ -118,6 +120,25 @@ InMemoryAnnData <- R6::R6Class("InMemoryAnnData", # nolint
...
+        private$.obsm
+      } else {
+        # TODO: validate obsm
+        private$.obsm <- private$.validate_array_collection_generic(value, "obsm", c(self$n_obs()), expected_rownames=rownames(self))
+        self
+      }
+    },
+    varm = function(value) {
+      if (missing(value)) {
+        private$.varm
```
Comment (rcannood) [inline R/InMemoryAnnData.R:136]: Add trackstatus comment

```suggestion
        # trackstatus: class=InMemoryAnnData, feature=get_varm, status=done
        private$.varm
```

File: R/HDF5AnnData.R [hunk 6]
```diff
@@ -36,6 +38,24 @@ HDF5AnnData <- R6::R6Class("HDF5AnnData", # nolint
         write_h5ad_element(value, private$.h5obj, "/layers")
       }
     },
+    #' @field obsm
+    obsm = function(value) {
+      if (missing(value)) {
+        read_h5ad_element(private$.h5obj, "obsm")
```
Comment (rcannood) [inline R/HDF5AnnData.R:44]: Add trackstatus comment

```suggestion
        # trackstatus: class=HDF5AnnData, feature=get_obsm, status=done
        read_h5ad_element(private$.h5obj, "obsm")
```

File: R/HDF5AnnData.R [hunk 7]
```diff
@@ -36,6 +38,24 @@ HDF5AnnData <- R6::R6Class("HDF5AnnData", # nolint
...
+      } else {
+        # TODO: validate obsm
+        write_h5ad_element(value, private$.h5obj, "/obsm")
+      }
+    },
+    #' @field varm
+    varm = function(value) {
+      if (missing(value)) {
+        read_h5ad_element(private$.h5obj, "varm")
```
Comment (rcannood) [inline R/HDF5AnnData.R:53]: Add trackstatus comment

```suggestion
        # trackstatus: class=HDF5AnnData, feature=get_varm, status=done
        read_h5ad_element(private$.h5obj, "varm")
```

File: R/HDF5AnnData.R [hunk 8]
```diff
@@ -36,6 +38,24 @@ HDF5AnnData <- R6::R6Class("HDF5AnnData", # nolint
         write_h5ad_element(value, private$.h5obj, "/layers")
       }
     },
+    #' @field obsm
+    obsm = function(value) {
+      if (missing(value)) {
+        read_h5ad_element(private$.h5obj, "obsm")
+      } else {
+        # TODO: validate obsm
```
Comment (rcannood) [inline R/HDF5AnnData.R:46]: Add trackstatus comment

```suggestion
        # trackstatus: class=HDF5AnnData, feature=set_obsm, status=wip
        # TODO: validate obsm
```

File: R/HDF5AnnData.R [hunk 9]
```diff
@@ -36,6 +38,24 @@ HDF5AnnData <- R6::R6Class("HDF5AnnData", # nolint
...
+        write_h5ad_element(value, private$.h5obj, "/obsm")
+      }
+    },
+    #' @field varm
+    varm = function(value) {
+      if (missing(value)) {
+        read_h5ad_element(private$.h5obj, "varm")
+      } else {
+        # TODO: validate varm
```
Comment (rcannood) [inline R/HDF5AnnData.R:55]: Add trackstatus comment

```suggestion
        # trackstatus: class=HDF5AnnData, feature=set_varm, status=wip
        # TODO: validate varm
```

**Review by rcannood — COMMENTED** [review]
Thanks for creating this PR and being involved in this project! :relaxed: 

I merged the main branch into this one, and I added a few suggestions to the code.

Could you run `styler::style_pkg()` to resolve some minor formatting issues?

> I think I would like to leave SingleCellExperiment + Seurat to separate PRs.

Makes sense to me!

File: R/AbstractAnnData.R [hunk 10]
```diff
@@ -155,6 +163,63 @@ AbstractAnnData <- R6::R6Class("AbstractAnnData", # nolint
...
+
+    # @description `.validate_array_generic()` checks that dimensions are
+    #   consistent with the anndata object.
+    # @param mat A matrix to validate
+    # @param label Must be `"X"` or `"layer[[...]]"` where `...` is
+    #   the name of a layer.
+    # @param shape Expected dimensions of matrix
+    # @param expected_rownames
+    # @param excepted_colnames
+    .validate_array_generic = function(mat, label, shape, expected_rownames = NULL, expected_colnames = NULL) {
```
Comment (ivirshup) [inline R/AbstractAnnData.R:175]: This could replace `.validate_layers`, should I do that in this PR?

Comment (rcannood) [reply R/AbstractAnnData.R:175]: Less code sounds good to me! ☺️

**Review by rcannood — APPROVED** [review]
LGTM! Might need to update the `trackstatus` comments since you added validation to obsm and varm.

**Pull request closed without merging**